### DISCUSSION TITLE
try catch 把我搞抑郁了

### DISCUSSION THREAD

**Discussion by leiganglee** [AmazingAng/WTF-Solidity#434] [Q&A] [answered]
Try Catch
try-catch可以捕获什么异常？  revert() require() assert()
以下异常返回值类型为bytes的是. assert()
try-catch捕获到异常后是否会使try-catch所在的方法调用失败. 不会
try代码块内的revert是否会被catch本身捕获  不会

上面的答案有哪个是错的吗。为什么总是提示我错了一个。这是哪个错了。

**Comment by chenso121** [comment]
怎么搞的吗，这个怎么提交都是错误的

**Reply by ranxi2001** [reply]
1.以上都可以
2.assert()
3.不会
4.不会

**Comment by chenso121** [comment]
谁能解答一下这个答案呀，卡死在这里。。。

**Reply by AmazingAng** [reply]
有没有可能不是这一题错了

**Reply by chenso121** [reply]
不懂，用时不够吗，前面的都没问题，还是尝试次数超限？真不知道错在哪里。。。

**Reply by AmazingAng** [reply]
没有时间和尝试限制的啊。。

**Reply by chenso121** [reply]
前两个没问题吧，1 以上都是，2 assert，4 不会 也没问题吧，能否告知哪里有问题

**Reply by chenso121** [reply]
看DC习题讨论上才知道，是不是题2答案有问题呀

**Reply by KrasCN** [reply]
以下异常返回值类型为bytes的是.   选以上都是的话 就能通过

**Reply by shaojunda** [reply]
多谢，就了命了，真的是，感觉第二题多少有点不严谨。

**Comment by YuanBoXie** [comment]
以下异常返回值类型为bytes的: 以上皆是

**Reply by trylovetom** [reply]
請問一下是題目2答案有問題, 還是 bytest 是標準答案?

**Comment by wiasliaw** [comment, marked as answer]
> 1. try-catch可以捕获什么异常？ 

只要能發出異常就能被捕捉，revert, require assert 都可以發出異常
catch 後面在乎的是 revert 回來的訊息能不能正確的被解碼

> 2. 以下异常返回值类型为bytes的是

revert, require assert 都是回傳 bytes，catch 後面在乎的是回傳的訊息能不能正確的被解碼


> 3. try-catch捕获到异常后是否会使try-catch所在的方法调用失败

有可能，catch 解碼失敗的異常沒有辦法被捕捉

> 4. try代码块内的revert是否会被catch本身捕获

不會


自己寫的有關 error 的文章：https://wiasliaw.io/articles/how-try-catch-works-in-solidity
